src/object/srv_obj_migrate.c

```diff
@@ -2070,6 +2119,150 @@ struct migrate_obj_val {
 	uint32_t	tgt_idx;
 };
 
+/* Destroys an object exactly one time per migration session. Uses the
+ * mpt_del_obj_tab field of the tls to store which object ids have already
+ * been deleted this session.
+ *
+ * Only used for reintegration
+ *
+ * Must be called on the appropriate VOS target xstream!
+ */
+static int
+destroy_existing_obj(struct migrate_pool_tls *tls, unsigned int tgt_idx,
+		     daos_unit_oid_t *oid, uuid_t cont_uuid)
+{
+	d_list_t *link;
+	int rc;
+
+	/* NOTE:
+	 * Note that only the object ID + shard is stored in the hash table, not
+	 * the padding that is also included in the daos_unit_oid_t struct
+	 * There will be many objects inserted in this hash table and so as
+	 * little data should be stored there as possible
+	 */
+	link = d_hash_rec_find(&tls->mpt_del_tabs[tgt_idx], oid,
+			       sizeof(daos_obj_id_t) + sizeof(uint32_t));
+	if (!link) {
+		/* Not actually storing anything in the table - just using it
+		 * to test set membership. The link stored is just the simplest
+		 * base list type
+		 */
+		struct migrate_del_obj_key *key;
+		struct ds_cont_child *cont;
+
+		rc = ds_cont_child_open_create(tls->mpt_pool_uuid,
+					       cont_uuid,
+					       &cont);
+		if (rc) {
+			D_WARN("Failed to open cont prior to migration; pool="
+			       DF_UUID" cont="DF_UUID"\n",
+			       DP_UUID(tls->mpt_pool_uuid), DP_UUID(cont_uuid));
+			/* This failure is not fatal - the object can't be
+			 * deleted, but it probably doesn't need to be if
+			 * the container can't even be opened. Proceed onwards,
+			 * marking the object as being ready to start migration
+			 */
+			D_GOTO(mark_obj_deleted, rc);
+		}
+
+		rc = vos_obj_delete(cont->sc_hdl, *oid);
+		if (rc != 0)
+			D_ERROR("Migrate failed to destroy object prior to "
+				"reintegration: pool/object "DF_UUID"/"DF_UOID
+				" rc: "DF_RC"\n",
+				DP_UUID(tls->mpt_pool_uuid), DP_UOID(*oid),
+				DP_RC(rc));
+			/*
+			 * Despite this error, continue anyway to add this key
+			 * to the table. Attempting to delete this object again
+			 * later will only make things worse
+			 */
+		else
+			D_DEBUG(DB_REBUILD,
+				"destroyed pool/object "DF_UUID"/"DF_UOID
+				" before reintegration\n",
+				DP_UUID(tls->mpt_pool_uuid), DP_UOID(*oid));
+
+		ds_cont_child_put(cont);
+
+mark_obj_deleted:
+
+		/* Insert a link into the hash table to mark this cont_uuid as
+		 * having already been initialized
+		 */
+		D_ALLOC_PTR(key);
+		if (key == NULL)
+			return -DER_NOMEM;
+
+		key->id_pub.hi = oid->id_pub.hi;
+		key->id_pub.lo = oid->id_pub.lo;
+		key->id_shard = oid->id_shard;
+
+		D_INIT_LIST_HEAD(&key->obj_link);
+		rc = d_hash_rec_insert(&tls->mpt_del_tabs[tgt_idx], key,
+				       sizeof(daos_obj_id_t) + sizeof(uint32_t),
+				       &key->obj_link, true);
```

Comment:
in current implementation, with these two trees  
        /* Container/objects already migrated will be attached to the tree, to
         * avoid the object being migrated multiple times.
         */
        daos_handle_t           mpt_migrated_root_hdl;
        struct btr_root         mpt_migrated_root;

The object should never be migrated twice, i.e. migrate_one_obj_prepare_ult should only be executed once for each object, so you do not need this extra tree here.

Reply:
Excellent observation - this simplifies the patch so ridiculously much. New patch will be up shortly, just testing now which removes this hash table.